doc/trader.6
.\" *********************************************************************
.\" *                                                                   *
.\" *           Star Traders: A Game of Interstellar Trading            *
.\" *              Copyright (C) 1990-2021, John Zaitseff               *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\" Author: John Zaitseff <J.Zaitseff@zap.org.au>
.\" $Id$
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or (at
.\" your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see https://www.gnu.org/licenses/.
.\"
.\"
.\" *********************************************************************
.\" Various macros
.\"
.\" Allow hyphenation
.hy 14
.\"
.\" Define .URL, .MTO and .FTP, then try to load the www.tmac version
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.de MTO
\\$2 \(la\\$1\(ra\\$3
..
.de FTP
\\$2 \(la\\$1\(ra\\$3
..
.if \n[.g] .mso www.tmac
.\"
.\" *********************************************************************
.TH TRADER 6 "19th January, 2021" "Unix-like systems"
.SH NAME
trader \- a game of interstellar trading
.\" *********************************************************************
.SH SYNOPSIS
.B trader
.RB [ \-\-no\-color | \-\-no\-colour ]
.RB [ \-\-max\-turn=\c
.IR NUM ]
.RI [ GAME ]
.br
.B trader
.RB [ \-h | \-\-help ]
.RB [ \-V | \-\-version ]
.\" *********************************************************************
.SH DESCRIPTION
\fBStar Traders\fP is a simple game of interstellar trading, where the
objective is to create companies, buy and sell shares, borrow and repay
money, in order to become the wealthiest player (the winner).
.\" *********************************************************************
.SH OPTIONS
.TP
.I GAME
If \fIGAME\fP is specified as a number between \fB1\fP and \fB9\fP
(inclusive), load and continue playing that game.  If \fIGAME\fP is not
specified, start a new game.
.TP
.BR \-\-no\-color ", " \-\-no\-colour
Don't use colour for displaying the text in the game.  Use this option
for a \*(lqretro-computing\*(rq look (as shown in the
.B EXAMPLES
below).
.TP
.BI \-\-max\-turn= NUM
Set the number of turns in the game to \fINUM\fP.  In this version of
Star Traders, \fINUM\fP must be greater or equal to 10.  If this option
is not specified, the default is 50 turns.
.TP
.BR \-h ", " \-\-help
Show a summary of command-line options and exit.
.TP
.BR \-V ", " \-\-version
Display version information about the program, then exit.
.\" *********************************************************************
.SH "EXIT STATUS"
.TP
.B 0
Star Traders finished without any errors.
.TP
.B 1
Star Traders encountered an unrecoverable problem or error; a diagnostic
message will be written to standard error in this case.
.\" *********************************************************************
.SH EXAMPLES
You can start a new game by running Star Traders without any command line
options:
.RS
.sp
.B trader
.sp
.RE
Once the game starts, you will be asked to enter the number of people
playing.  From one to eight people can play (although, in this version,
they will all have to share the one keyboard and screen!).  After
entering the names of the players, you will have the opportunity to read
instructions on how to play the game.  Do so\(emand good luck in the
game!
.PP
If you would like to continue a previously-saved game, simply specify
that game number on the command line.  For example, the following starts
game 4, if it was previously saved:
.RS
.sp
.B trader 4
.sp
.RE
If you are running under the X Window System, you might like to start the
game in a dedicated
.BR xterm (1)
window (typed all on one line):
.RS
.sp
.nh
.na
.B "xterm \-g 80x24 \-fa Mono \-fs 18 \-bg black \-fg white \-bc +sb +fbx \-e trader &"
.ad
.hy 14
.sp
.RE
If you would like a full \*(lqretro-computing\*(rq green-screen
experience, try (again, typed all on one line):
.RS
.sp
.nh
.na
.B "xterm \-g 80x24 \-fa Mono \-fs 18 \-bg \(aq#181818\(aq \-fg \(aq#2CAB00\(aq \-bc +sb +bdc +fbx \-xrm \(aqXTerm*colorBD: #41FF00\(aq \-e trader \-\-no\-colour &"
.ad
.hy 14
.sp
.RE
Or, if you prefer the old amber screens of yesteryear:
.RS
.sp
.nh
.na
.B "xterm \-g 80x24 \-fa Mono \-fs 18 \-bg \(aq#101010\(aq \-fg \(aq#AB7A00\(aq \-bc +sb +bdc +fbx \-xrm \(aqXTerm*colorBD: #FFB700\(aq \-e trader \-\-no\-colour &"
.ad
.hy 14
.sp
.RE
.\" *********************************************************************
.SH ENVIRONMENT
.TP
.BR XDG_DATA_HOME ", " HOME
If \fBXDG_DATA_HOME\fP is set to an absolute pathname (that is, a path
that starts with \*(lq\fI/\fP\*(rq), Star Traders will use that
directory, with a subdirectory \fItrader\fP, to store game files.  If
this environment variable is not set or does not start with
\*(lq\fI/\fP\*(rq, \fI\(ti/.local/share/trader\fP will be used instead,
where \*(lq\fI\(ti\fP\*(rq represents your home directory, as contained
in the \fBHOME\fP environment variable.
.TP
.BR LINES ", " COLUMNS
Star Traders uses the Curses library for displaying text on the screen.
As such, it will access these two environment variables if the underlying
Curses library does so (see, for example, the \fBENVIRONMENT\fP section
in the
.BR ncurses (3)
manual page for in-depth details).  It requires a text console or window
of at least 80\(mu24 in size.
.TP
.BR LANG ", " LC_ALL ", etc."
This version of Star Traders has full support for locales and will use
appropriate settings.  In particular, messages will be displayed using
\fBLC_MESSAGES\fP and \fBLANGUAGE\fP (if Star Traders has been translated
into that language).  In addition, numeric quantities will be displayed
using \fBLC_NUMERIC\fP and monetary quantities will use
\fBLC_MONETARY\fP.  See the
.BR locale (7)
or
.BR setlocale (3)
manual pages for more details on locale settings.
.TP
.B TEXTDOMAINDIR
If set, Star Traders will use this path as the base with which to locate
its message catalogs instead of the compiled-in path; the relevant
\fItrader.mo\fP files should be located in language-code subdirectories
(such as \fIen_AU\fP), in \fILC_MESSAGES\fP sub-subdirectories.
.\" *********************************************************************
.SH FILES
.TP
.IB \(ti/.local/share/trader/game N
Star Traders stores saved game files in the \fI.local/share/trader\fP
subdirectory in your home directory (unless overriden by the
\fBXDG_DATA_HOME\fP environment variable).  \fIN\fP is a number between
\fB1\fP and \fB9\fP inclusive.  The game file is scrambled to prevent you
or others from casually cheating!
.TP
.IB \(ti/.trader/game N
If the \fI\(ti/.trader\fP directory exists, game files will be read from
and saved to this location instead.  This is for compatibility with
versions of Star Traders prior to version 7.15.
.\" *********************************************************************
.SH BUGS
None yet known...
.\" *********************************************************************
.SH FEEDBACK
Your comments, suggestions, corrections and enhancements are always
warmly welcomed!  Please send these to:
.IP Postal: 10n
\fBJohn Zaitseff\fP,
.br
The ZAP Group,
.br
Unit 6, 116 Woodburn Road,
.br
Berala, NSW, 2141,
.br
Australia
.IP Email: 10n
.MTO J.Zaitseff@zap.org.au
.PD 0
.IP Web: 10n
.URL https://www.zap.org.au/projects/trader/
.IP FTP: 10n
.URL https://ftp.zap.org.au/pub/trader/
.br
.FTP ftp://ftp.zap.org.au/pub/trader/
.PD 0
.IP Git: 10n
.URL https://www.zap.org.au/git-browser/trader.git
.br
.URL https://git.zap.org.au/git/trader.git
.br
.URL git://git.zap.org.au/data/git/trader.git
.PD
.PP
.\" *********************************************************************
.SH COPYRIGHT
Copyright \(co 1990\-2021, John Zaitseff.
.PP
\fBStar Traders\fP is free software that is distributed under the terms
of the GNU General Public License.  You can redistribute it and/or modify
it under the terms of that License as published by the Free Software
Foundation, either version 3 or (at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program.  If not, see the
.URL https://www.gnu.org/licenses/ "GNU licenses web page" .
.PP
Even though the GNU General Public License does \fInot\fP require you to
send your modifications back to the author, it is considered \*(lqgood
form\*(rq to do so, as this allows your modifications to be incorporated
into future versions of the program, allowing others to benefit from
them.
.\" *********************************************************************
.SH HISTORY
The original (and very primitive) Star Traders game was written by
S.\~J\. Singer in 1984 using Altair Basic.  This was modified for
Microsoft Basic (MBASIC) running under the CP/M-80 operating system by
John Zaitseff and released on 7th March, 1988.
.PP
Star Traders was then completely rewritten in 1990 for the
Australian-designed 8-bit MicroBee computer running CP/M-80 on a Zilog
Z80 processor, using Turbo Pascal 3.01a.  Essentially, only the name of
the game and some of the ideas were retained in this version.  Version
4.1 of Star Traders was released on 1st August, 1991.
.PP
In 1992, it was recompiled for the NEC Advanced Personal Computer (with
8-inch floppy drives!) running CP/M-86 on an 8086 processor, using Turbo
Pascal 2.0.  This version had colour added to it in the form of ANSI
escape sequences; version 4.4 was released on 2nd August, 1993.
.PP
The next version came in 1993, when the program was recompiled to run on
IBM-compatible machines running MS-DOS and ANSI.SYS.  Turbo Pascal 6.0
was used for this.  The ANSI escape sequences were slightly different
under MS-DOS than under the NEC, in that the NEC supported a number of
extra character attributes.  In other words, the MS-DOS version looked
worse than the one running under CP/M-86!
.PP
Star Traders was recompiled again in 1994 for IBM-compatible machines
with VGA/EGA/CGA video graphics adapters.  The output routines were
recoded to use a \*(lqwindowed\*(rq look.  Borland Pascal 7.0 was used
for this purpose, along with a number of text window manipulation
modules.  Version 5.4 was released on 1st June, 1994.
.PP
In 1995, Star Traders was completely rewritten for the 16-bit Microsoft
Windows 3.1 graphical environment.  Microsoft Visual Basic 3.0 was used
for this purpose.  Although completely rewritten, the original algorithms
were reused from previous versions.  Version 6.0 of the game was released
on 15th September, 1995.
.PP
Star Traders was then to languish until almost 16 years later... when the
game was rewritten once again, this time in the C programming language.
Version 7.0 was released on 25th July, 2011 for Unix-like operating
systems such as Linux, with subsequent releases to add features and
correct bugs.  Now you, too, can run this small piece of computing
history!
.\" *********************************************************************
.SH "SEE ALSO"
.URL https://www.zap.org.au/projects/trader/ "Star Traders home page"
.\" *********************************************************************
.\" End of file


NEWS
.. -*- mode: rst; coding: utf-8 -*-
.. role:: filename(literal)
.. default-role:: filename

============================================
Star Traders: A Game of Interstellar Trading
============================================

Star Traders is a simple game of interstellar trading, where the
objective is to create companies, buy and sell shares, borrow and repay
money, in order to become the wealthiest player (the winner).

The following history is presented in reverse chronological order.  If
you would like to see a list of detailed changes to this project, you
should browse the `Star Traders Git repository`__ on `The ZAP Group web
server`__.

__ https://www.zap.org.au/git-browser/trader.git
__ https://www.zap.org.au/


.. contents::


Version 7.16 (19th January, 2021)
---------------------------------

Added code to allow the ``TEXTDOMAINDIR`` environment variable to specify
where compiled translations are located.  Updated the documentation to
mention this new feature.

Added an application metadata file that conforms to the AppStream 0.12
specification.  Updated all translations to suit, using parts of existing
strings.

Updated the German translation, with thanks to Mario Blättermann.


Version 7.15 (12th January, 2021)
---------------------------------

Changed the default location where game files are saved from `~/.trader`
to `~/.local/share/trader`, so as to follow the `XDG Base Directory
Specification`__.  If set to an absolute path, the ``XDG_DATA_HOME``
environment variable will override this location.  If the `~/.trader`
directory is already present, it will continue to be used.  Updated the
manual page to suit.

__ https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

Updated the Russian translation to use the uppercase polite singular
forms «Вы» and «Ваш», as is customary in written documents.

Updated to GNU Gettext 0.20 or later.  Updated to the latest snapshot of
the Gnulib GNU Portability Library.  Also updated the `INSTALL` file to
list the latest tested versions of operating systems, including a new
listing for Haiku R1/beta2.


Version 7.14 (9th January, 2020)
--------------------------------

Added the Brazilian Portuguese translation, with thanks to Rafael
Fontenelle.  Updated the French, Hungarian and Esperanto translations, in
that order, with thanks to Frédéric Marchal, Balázs Úr and Felipe Castro.

Updated to the latest snapshot of the Gnulib GNU Portability Library.
Also updated the `INSTALL` file to list macOS 10.15 (Catalina) and the
64-bit version of Cygwin under Microsoft Windows 10 as being tested
platforms.


Version 7.13 (13th November, 2019)
----------------------------------

Converted all text files to reStructuredText (ReST) format.  This format
is still quite readable as raw text, and is much more capable than the
myriad all-slightly-incompatible variants of Markdown.

Modified the source code to work more reliably on non-GNU C Library
(``glibc``) platforms, particularly FreeBSD, Solaris and Cygwin, when
using the standard POSIX locale.  Extensively tested the build procedure
on multiple operating systems and compilers, and submitted code upstream
to the Autoconf Archive project to make this work more robustly.  Listed
tested operating system and compiler combinations in the `INSTALL` file.

Updated the Esperanto translation, with thanks to Felipe Castro.

Updated one string in the game for consistency with other labels; updated
all PO translations to suit.  Also added strings from obsolete versions
of the GNU C Library so that Star Traders can provide translations for
them.

Updated to the latest snapshot of the Gnulib GNU Portability Library.  In
addition, updated the list of prerequisites in the `INSTALL` file and
listed some more command line options for the ``configure`` script.


Version 7.12 (24th August, 2017)
--------------------------------

Updated the Serbian translation, with thanks to Мирослав Николић.


Version 7.11 (18th June, 2017)
------------------------------

Bug fix: some architectures (such as ARM, PowerPC and S390) treat
:code:`char` as :code:`unsigned char`, and this caused
:code:`b64decode()` to fail while loading a saved game.  Changed the
relevant type from :code:`char` to :code:`int` to fix this.


Version 7.10 (2nd June, 2017)
-----------------------------

Added a `data` directory that contains a desktop file and appropriate
icons.  Adjusted the game constants for more dynamic play: allow the
share prices and returns to change a bit more frequently and more
rapidly.  Tweaked the colours in the example monochrome retro-computing
command lines, including adding an amber monitor version.

Added a Serbian translation, with thanks to Мирослав Николић.  Updated
the Danish, French, Swedish, Hungarian, Norwegian Bokmål and German
translations, in that order, with thanks to Joe Hansen, Frédéric Marchal,
Anders Jonsson, Balázs Úr, Johnny A. Solbu and Philipp Thomas
respectively.

Updated translatable strings to overcome a limitation of ``msgfmt``\(1):
if a string starts or ends with a new-line character, its translation
must do so as well.  This has an impact on labels in particular.  Also
changed the help text string “Page *N* of *M*” to allow translations of
the form “Of *M* pages, this is page *N*”, and all quoted strings using
the obsolete GNU form ```like this'`` to ``'this'``.  In addition, newer
versions of `lib/getopt.c` have different strings.  All translations have
been updated as a result.

Removed the obsolete Gnulib ``gettext`` module and updated to the latest
Git snapshot of Gnulib.  Updated the dependency version numbers for
Autoconf, Automake and Gettext.  Also updated all macros from the
Autoconf Archive.

Migrated the source code repository from Subversion to Git.  Updated the
documentation to suit, and also listed specific required version numbers
for the Autoconf and Automake tools in the `INSTALL` file.


Version 7.9 (5th January, 2016)
-------------------------------


src/trader.c
/************************************************************************
*                                                                       *
*             Star Traders: A Game of Interstellar Trading              *
*                Copyright (C) 1990-2021, John Zaitseff                 *
*                                                                       *
************************************************************************/

/*
  Author: John Zaitseff <J.Zaitseff@zap.org.au>
  $Id$

  Star Traders is a simple game of interstellar trading, where the object
  of the game is to create companies, buy and sell shares, borrow and
  repay money, in order to become the wealthiest player (the winner).

  This file, trader.c, contains the main program and command-line
  interface for Star Traders.


  This program is free software: you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by the
  Free Software Foundation, either version 3 of the License, or (at your
  option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see https://www.gnu.org/licenses/.
*/


#include "trader.h"


/************************************************************************
*                 Module-specific constant definitions                  *
************************************************************************/

// Constants for command line options

enum options_char {
    OPTION_NO_COLOR = 1,
    OPTION_DONT_ENCRYPT,
    OPTION_MAX_TURN
};

static const char options_short[] = "hV";
    // -h, --help
    // -V, --version

static struct option const options_long[] = {
    { "help",         no_argument,       NULL, 'h' },
    { "version",      no_argument,       NULL, 'V' },
    { "no-color",     no_argument,       NULL, OPTION_NO_COLOR },
    { "no-colour",    no_argument,       NULL, OPTION_NO_COLOR },
    { "dont-encrypt", no_argument,       NULL, OPTION_DONT_ENCRYPT },
    { "max-turn",     required_argument, NULL, OPTION_MAX_TURN },
    { NULL,           0,                 NULL, 0 }
};


/************************************************************************
*                  Module-specific function prototypes                  *
************************************************************************/

/*
  Function:   main - Main program implementing Star Traders
  Parameters: argc - Command-line argument count
              argv - Command-line argument vector
  Returns:    int  - Operating system return code: 0 if all well, 1 if not.

  The main() function is, of course, where all the action starts in a C
  program.  This function contains the main game loop, a series of game
  functions that are called in sequence.
*/
int main (int argc, char *argv[]);


/*
  Function:   process_cmdline - Process command line arguments
  Parameters: argc            - Same as passed to main()
              argv            - Same as passed to main()
  Returns:    (nothing)

  This function processes the command line arguments passed through argc
  and argv.  If required, it shows the program version number and/or
  command-line help.  It also sets global variables starting with option_
  to appropriate values.
*/
static void process_cmdline (int argc, char *argv[]);


/*
  Function:   show_version - Show program version information
  Parameters: (none)
  Returns:    (does not return)

  This function displays version information about this program, then
  terminates with exit code EXIT_SUCCESS.
*/
static void show_version (void) __attribute__((noreturn));


/*
  Function:   show_usage - Show command line usage information
  Parameters: status     - Exit status
  Returns:    (does not return)

  This function displays usage information for this program.  If status
  is zero, a detailed explanation is sent to stdout; otherwise, a brief
  message is sent to stderr.  It exits to the operating system with
  status as the exit code.
*/
static void show_usage (int status) __attribute__((noreturn));


/*
  Function:   init_program_prelim - Initialise program preliminaries
  Parameters: argc                - Same as passed to main()
              argv                - Same as passed to main()
  Returns:    (nothing)

  This function initialises the program name, locale and message
  catalogs.  It must be called before command line arguments are
  processed.
*/
static void init_program_prelim (int argc, char *argv[]);


/*
  Function:   init_program - Initialise program-wide functions
  Parameters: (none)
  Returns:    (nothing)

  This function initialises the terminal display, internal low-level
  routines, etc.  It should be called before the game starts.
*/
static void init_program (void);


/*
  Function:   end_program - Deinitialise program-wide functions
  Parameters: (none)
  Returns:    (nothing)

  This function finalises the terminal display, internal low-level
  routines, etc.  It should be the last function called in the ordinary
  course of program execution.
*/
static void end_program (void);


/************************************************************************
*                             Main program                              *
************************************************************************/

int main (int argc, char *argv[])
{
    // Initialise program name, locale and message catalogs
    init_program_prelim(argc, argv);

    // Process command line arguments
    process_cmdline(argc, argv);

    // Set up the display, internal low-level routines, etc.
    init_program();

    // Play the actual game
    init_game();
    while (! quit_selected && ! abort_game && turn_number <= max_turn) {
	selection_t selection;

	select_moves();
	selection = get_move();
	process_move(selection);
	exchange_stock();
	next_player();
    }
    end_game();

    // Finish up...
    end_program();
    return EXIT_SUCCESS;
}


/************************************************************************
*                        Command line processing                        *
************************************************************************/

// These functions are documented at the start of this file


/***********************************************************************/
// process_cmdline: Process command line arguments

void process_cmdline (int argc, char *argv[])
{
    // Process arguments starting with "-" or "--"
    opterr = true;
    while (true) {
	int c = getopt_long(argc, argv, options_short, options_long, NULL);
	if (c == EOF)
	    break;

	switch (c) {
	case 'h':
	    // -h, --help: show help
	    show_usage(EXIT_SUCCESS);
	    break;

	case 'V':
	    // -V, --version: show version information
	    show_version();
	    break;

	case OPTION_NO_COLOR:
	    // --no-color, --no-colour: don't use colour
	    option_no_color = true;
	    break;

	case OPTION_DONT_ENCRYPT:
	    // --dont-encrypt: don't encrypt game files
	    option_dont_encrypt = true;
	    break;

	case OPTION_MAX_TURN:
	    // --max-turn: specify the maximum turn number
	    {
		char *p;

		option_max_turn = strtol(optarg, &p, 10);

		if (option_max_turn < MIN_MAX_TURN || p == NULL || *p != '\0') {
		    fprintf(stderr, _("%s: invalid value for --max-turn: '%s'\n"),
			    program_name, optarg);
		    show_usage(EXIT_FAILURE);
		}
	    }
	    break;

	default:
	    show_usage(EXIT_FAILURE);
	}
    }

    // Process remaining arguments

    if (optind < argc && argv[optind] != NULL) {
	if (*argv[optind] == '-') {
	    fprintf(stderr, _("%s: invalid operand '%s'\n"),
		    program_name, argv[optind]);
	    show_usage(EXIT_FAILURE);
	}

	if (strlen(argv[optind]) == 1
	    && *argv[optind] >= '1' && *argv[optind] <= '9') {
	    game_num = *argv[optind] - '0';
	} else {
	    fprintf(stderr, _("%s: invalid game number '%s'\n"),
		    program_name, argv[optind]);
	    show_usage(EXIT_FAILURE);
	}

	optind++;
    }

    if (optind < argc && argv[optind] != NULL) {
	fprintf(stderr, _("%s: extra operand '%s'\n"),
		program_name, argv[optind]);
	show_usage(EXIT_FAILURE);
    }
}


/***********************************************************************/
// show_version: Show program version information

void show_version (void)
{
    /* TRANSLATORS: "John Zaitseff" [IPA d͡ʒɒn ˈzaɪ̯t͡səf] is the proper
       name of the author.  The IPA pronunciation in this comment is in
       UTF-8 encoding. */
    printf(_("\
Star Traders (%s) %s\n\
Copyright (C) %s, John Zaitseff.\n\
\n\
Star Traders is a simple game of interstellar trading, where the object\n\
of the game is to create companies, buy and sell shares, borrow and repay\n\
money, in order to become the wealthiest player (the winner).\n\
\n\
This program is free software that is distributed under the terms of the\n\
GNU General Public License, version 3 or later.  You are welcome to\n\
modify and/or distribute it under certain conditions.  This program has\n\
NO WARRANTY, to the extent permitted by law; see the License for details.\n\
"), program_name, PACKAGE_VERSION, "1990-2021");

    exit(EXIT_SUCCESS);
}


/***********************************************************************/
// show_usage: Show command line usage information

void show_usage (int status)
{
    if (status != EXIT_SUCCESS) {
	fprintf(stderr, _("%s: Try '%s --help' for more information.\n"),
		program_name, program_name);
    } else {
	printf(_("Usage: %s [OPTION ...] [GAME]\n"), program_name);
	printf(_("\
Play Star Traders, a simple game of interstellar trading.\n\n\
"));
	printf(_("\
Options:\n\
  -V, --version        output version information and exit\n\
  -h, --help           display this help and exit\n\
      --no-color       don't use color for displaying text\n\
      --max-turn=NUM   set the number of turns to NUM\n\n\
"));
	printf(_("\
If GAME is specified as a number between 1 and 9, load and continue\n\
playing that game.  If GAME is not specified, start a new game.\n\n\
"));

#ifdef PACKAGE_AUTHOR
	/* TRANSLATORS: The first %s is the proper name of the package
	   author, John Zaitseff [IPA d͡ʒɒn ˈzaɪ̯t͡səf]; the second %s is
	   the email address for reporting bugs.  Please add ANOTHER line
	   with the (translated) text "Report translation bugs to
	   <ADDRESS>\n", with ADDRESS replaced with either an email
	   address or web URL for reporting bugs in your translation. */
	printf(_("Report bugs to %s <%s>.\n"), PACKAGE_AUTHOR, PACKAGE_BUGREPORT);
#else
	/* TRANSLATORS: %s is the email address for reporting bugs.  As
	   with the previous string, please add ANOTHER line with the
	   (translated) text "Report translation bugs to <ADDRESS>\n",
	   with ADDRESS replaced with either an email address or web URL
	   for reporting bugs in your translation. */
	printf(_("Report bugs to <%s>.\n"), PACKAGE_BUGREPORT);
#endif
#ifdef PACKAGE_PACKAGER_BUG_REPORTS
	/* TRANSLATORS: The first %s is for packagers and may be
	   something like "Debian". */
	printf(_("Report %s bugs to <%s>.\n"), PACKAGE_PACKAGER, PACKAGE_PACKAGER_BUG_REPORTS);
#endif
#ifdef PACKAGE_URL
	printf(_("Star Traders home page: <%s>.\n"), PACKAGE_URL);
#endif
    }

    exit(status);
}


/************************************************************************
*       Initialisation and deinitialisation function definitions        *
************************************************************************/

// These functions are documented at the start of this file


/***********************************************************************/
// init_program_prelim: Initialise program preliminaries

void init_program_prelim (int argc, char *argv[])
{
    // Strip off leading pathname components from program name
    init_program_name(argv[0]);

    // Initialise the locale and message catalogs
    init_locale();
}


/***********************************************************************/
// init_program: Initialise program-wide functions

void init_program (void)
{
    // Initialise the random number generator
    init_rand();

    // Initialise locale-specific variables
    init_locale_vars();

    // Initialise the terminal display
    init_screen();
}


/***********************************************************************/
// end_program: Deinitialise program-wide functions

void end_program (void)
{
    end_screen();
}


/***********************************************************************/
// End of file
